include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,512 @@
+#ifndef RBTREE_HEADER_H
+#define RBTREE_HEADER_H
+
+/* Adapted from https://github.com/BartoszMilewski/Okasaki/tree/master/RBTree */
+/* Implementing algorithms from                                               *
+ * https://matt.might.net/papers/germane2014deletion.pdf                      */
+
+#include <cassert>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#define CONSTRUCT_MSG_AND_THROW(msg)                                           \
+  do {                                                                         \
+    std::string emsg = std::string("[") + std::string(__func__)                \
+                       + std::string("] ") + std::string(msg);                 \
+    throw std::invalid_argument(emsg);                                         \
+  } while (false)
+
+//---               Ordered map on top of a red-black tree.                ---//
+
+// 1. No red node has a red child.
+// 2. Every path from root to empty node contains the same
+// number of black nodes.
+
+template <class T, class V>
+class RBTree {
+  enum Color { R, B, BB };
+
+  static Color minusOneColor(Color c) {
+    switch (c) {
+    case B: return R;
+    case BB: return B;
+    default: CONSTRUCT_MSG_AND_THROW("Unexpected color");
+    }
+  }
+
+  struct Node {
+    Node(Color c)
+        : _c(c)
+        , _s(0) { }
+    Color _c;
+    size_t _s;
+
+    virtual bool isLeaf() const = 0;
+    virtual ~Node(){};
+  };
+
+  struct Leaf : public Node {
+    Leaf(Color c)
+        : Node(c) {
+      assert(c == B || c == BB);
+    }
+    virtual bool isLeaf() const override { return true; }
+    virtual ~Leaf(){};
+  };
+
+  struct InternalNode : public Node {
+    InternalNode(
+        Color c, std::shared_ptr<const Node> const &lft, T key, V val,
+        std::shared_ptr<const Node> const &rgt)
+        : Node(c)
+        , _lft(lft)
+        , _key(key)
+        , _val(val)
+        , _rgt(rgt) {
+      this->_s = 1 + _lft->_s + _rgt->_s;
+    }
+    std::shared_ptr<const Node> _lft;
+    T _key;
+    V _val;
+    std::shared_ptr<const Node> _rgt;
+
+    virtual bool isLeaf() const override { return false; }
+    virtual ~InternalNode(){};
+  };
+
+  explicit RBTree(std::shared_ptr<const Node> const &node)
+      : _root(node) { }
+
+  Color rootColor() const {
+    assert(!isEmpty());
+    return _root->_c;
+  }
+
+  Color leafColor() const {
+    assert(isEmpty());
+    return _root->_c;
+  }
+
+public:
+  RBTree()
```

Comment:
all of these public operations need documentation describing what they do and what arguments they take.

Reply:
Added more documentation, specifically to public methods.